int32[] botIds
trajectory[] trajectories